feat: Insert Missing Comma on Enter

AgentRBY opened the pull request:
Fix #5 

P.s.: I have not written extensions in VS Code before, so I am open to suggestions

AgentRBY committed:
Add "Insert Comma"

AgentRBY committed:
Update docs

AgentRBY committed:
Fix typo

zardoy commented on package.json:
```diff
@@ -67,6 +67,11 @@
                     "type": "boolean",
                     "default": true
+                },
+                "insertCommaAfterEnter": {
```
```suggestion
                "insertMissingCommaOnEnter": {
```
As I remember in WS the feature is called `Insert missing commas on Enter` let's call it in the same way

zardoy commented on src/extension.ts:
```diff
@@ -110,4 +112,81 @@ export const activate = () => {
+            }
+
+            if (
+                !content.text.startsWith("\n") &&
```
Why don't we use `/^\n/r?$/.test(content.text)` instead of `startsWith`? I mean are there any cases where we need startsWith?

AgentRBY replied:
The regex is very slow, so I try not to use it where possible

zardoy commented:
Wow looks interesting! I think current implementation is fine enough, but probably only for json without comments. Initially I was thinking of using some parsing method to insert comma more precisely, some cases to test:

I marked with `|` positions where comma would be inserted incorrectly.
```jsonc
{
  "key": "value" /* comment */ // case 1|
}
```

Btw I checked that WebStorm doesn't do this auto comma inserting after comments and don't think its right.

I think we can easily workaround this comments problem just by proceeding the editor text with `strip-json-comments` module.

Btw I see I didn't check this project for a long time, so I'll try to setup a decent CI soon.

zardoy commented on package.json:
```diff
@@ -3,7 +3,7 @@
     "description": "Fix missing and trailing commas on save",
     "publisher": "zardoy",
-    "version": "0.0.0-dev",
+    "version": "0.0.1",
```
```suggestion
    "version": "0.0.0-dev",
```

Version managed automatically, revert this one. The version [is set](https://github.com/zardoy/release-action/blob/b877a561231a199a7c48e849b8561196f5272207/src/library/main.ts#L112) at publishing.

This is actually a common practice for guys using [semantic-release like tools](https://github.com/semantic-release/semantic-release/blob/27b105337b16dfdffb0dfa36d1178015e7ba68a3/package.json#L4)

zardoy commented on src/extension.ts:
```diff
@@ -110,4 +112,81 @@ export const activate = () => {
+                return;
+            }
+
+            const content = contentChanges[0];
```
Also let's support multicuror editing. Wrap code below into `editor.edit()` and then into a loop. Like so:

```ts
void editor.edit((edit) => {
    for (const contentChange of contentChanges) {
        // basically the same code as below but with continue instead of return:
        // do edit.insert if needed on specific contentChange
    }
})
```

AgentRBY committed:
Back version

AgentRBY committed:
Rename to insertMissingCommaOnEnter

AgentRBY committed:
Advanced support JSONC

AgentRBY committed:
Add multicursor support

AgentRBY commented:
Thx for feedback!
I fixed everyting exept `startsWith` because regex is very slow.


> I think we can easily workaround this comments problem just by proceeding the editor text with `strip-json-comments` module.

Also added JSONC support with "strip-json-comments"


In multicursor mode i have some problems:
1. prevLinePosition did not point to the last character, so I create a new Position
2. In some cases the comma is not inserted (see gif) and IDK how fix that 
![Code_KY8XwsBlf1](https://user-images.githubusercontent.com/68181944/202897972-c69b715c-a852-47aa-b520-e410b7f887e6.gif)

zardoy commented on src/extension.ts:
```diff
@@ -110,4 +113,89 @@ export const activate = () => {
+            }
+
+            editor.edit((edit) => {
+                contentChanges.forEach(async (content) => {
```
```suggestion
                contentChanges.forEach(async (change) => {
```

I know its just a naming, but current name `content` can lead to misunderstanding (it doesn't hold any actual content and its not `string` type)

zardoy committed:
fix eslint issues

zardoy commented on src/extension.ts:
```diff
@@ -110,4 +113,89 @@ export const activate = () => {
+            editor.edit((edit) => {
+                contentChanges.forEach(async (content) => {
+                    if (
+                        !content.text.startsWith("\n") &&
```
I still don't get the idea of using `startsWith` here. We're doing it only after enter keypress, right? So why don't we use `===` then? It also catches multiline pasting, but why do we need it?

zardoy replied:
Also let's move this check to in outer scope e.g.

```ts
if (contentChanges.some(change => /* check whether not enter keypress change */) return
```

I think this is still valid since we want to do this only after simple Enter keypress. If some edits adding newlines and some not, then most probably changes were done by some extension and not user.

AgentRBY replied:
> I still don't get the idea of using `startsWith` here. We're doing it only after enter keypress, right? So why don't we use `===` then? It also catches multiline pasting, but why do we need it?

After pressing `Enter` the value of `content.text` will be equal to `\n\r` (or `\n`, depending on the EoL type) plus the spaces following it. If we use trim(), it will remove not only spaces, but also the string EoL itself. So the only option I found is to check with `startsWith`.

zardoy replied:
> After pressing Enter the value of content.text will be equal to \n\r (or \n, depending on the EoL type) plus the spaces following it. 

Makes sense, but why did you switch to `.includes()` then? Anyway why don't we use the following code to be strict:

```ts
const firstTextChange = contentChanges[1]!.text
if (contentChanges.some((change) => change.text !== firstTextChange) || !/^\n\r? *$/.test(firstTextChange)) {
    return
}
```

AgentRBY replied:
> Makes sense, but why did you switch to .includes() then? 

Because in this situation there is no difference between `includes` and `startsWith`, and I didn't want to use the name startsWithEoL for function because it doesn't sound good 😄

>Anyway why don't we use the following code to be strict:

I don't really understand what this code does. And, again, I don't really want to use regex

zardoy replied:
> I don't really understand what this code does. And, again, I don't really want to use regex

Okay, I see, there is a simpler solution without regexs then:

```ts
const firstTextChange = contentChanges[0]!.text
const firstChangeIsEnterOnly = firstTextChange.startsWith('\n') && firstTextChange.trim() === ''
const changesAreTheSame = contentChanges.some((change) => change.text !== firstTextChange)
if (!changesAreTheSame || !firstChangeIsEnterOnly) {
    return
}

```

@AgentRBY What do you think now?

zardoy replied:
This is important as there is no way to determine wether user did Enter keypress or content pasting so we need to ensure we don't run it when user pasted content that has newline somewhere in the middle.

AgentRBY replied:
> This is important as there is no way to determine wether user did Enter keypress or content pasting so we need to ensure we don't run it when user pasted content that has newline somewhere in the middle.

Can you tell me a case where your current code doesn't work? 
I have long ago added a check to see if the current line is empty

![image](https://user-images.githubusercontent.com/68181944/203377043-39766a8a-05fc-46e1-b969-0c7f01be3148.png)

And it works quite well

![Code_J7Pe1vL7hX](https://user-images.githubusercontent.com/68181944/203377313-038ca189-80e9-4092-863e-616b3256c74a.gif)

zardoy replied:
I believe you didn't get me right. I was talking about content pasting (when you do ctrl+V the pasted content will be contentChanges`, as you probably mentioned in https://github.com/zardoy/vscode-fix-all-json/pull/6#issuecomment-1324049348). Again,
> content pasting so we need to ensure we don't run it when user pasted content that has newline somewhere in the middle

I want to exclude potential trigger from content pasting (https://github.com/zardoy/vscode-fix-all-json/pull/6#issuecomment-1324049348). Also, the code I provided to check content change `firstTextChange.startsWith('\n') && firstTextChange.trim() === ''` is more explicit rather than `.includes('\n')`

AgentRBY replied:
I added your check to the last commit. Please check it @zardoy

zardoy replied:
Thanks!

zardoy commented on src/utils.ts:
```diff
@@ -0,0 +1,7 @@
+}
+
+export function isNumber(text: string) {
+    return !Number.isNaN(Number(text));
```
```suggestion
    return !Number.isNaN(Number.parseInt(text, 10));
```

` ` (string with whitespace) char would be parsed to `0`.

Just finally checked implementation within vscode, this is incorrect way to do the parsing (have no idea why my linter didn't catch it).

zardoy committed:
simplify code

zardoy commented on src/extension.ts:
```diff
@@ -110,4 +113,89 @@ export const activate = () => {
+                        isNumber(prevLineLastChar);
+
+                    if (isMatchValue && isCurrentLineEmpty) {
+                        // In multicursor mode last character position is broken when I use prevLinePosition, IDK why
```
I faced similar probleim in
https://github.com/zardoy/vscode-better-snippets/blob/2ce9b402d94759cfe795d0f087dcfe5134b5523e/src/extension.ts#L391

This is because here, you work with **already changed document** (`document.lineAt`). For example you have the following document (| - cursor position):

```
1. {
2.     "test1": "a"|
3.     "test2": "a"|
4. }
```

After pressing enter:

```
1. {
2.    "test1": "a"
3.    |
4.    "test2": "a"
5.    |
6. }
```

In this callback you receive information that lines `2` and `3` has typed `\n` character. And now you can guess what line content you receive at the end (`"` and ` `). Also you were working with unsorted contentChanges so most probably code was trying first to get 3rd line and only then 2nd.

> Note: I fixed it with last commit, do you mind testing it?

AgentRBY replied:
Tested, works very well, thanks for the help

![Code_6ZZi5G5vds](https://user-images.githubusercontent.com/68181944/203049286-ee7bd1ab-4d43-4788-975a-432c37a29513.gif)

zardoy committed:
fix incorrect multicursor insertion

zardoy commented on src/extension.ts:
```diff
@@ -110,4 +113,89 @@ export const activate = () => {
+                        edit.insert(lastCharacterPosition, ",");
+                    }
+                })
+            });
```
What do you think of adding

```ts
{
    undoStopAfter: false,
    undoStopBefore: false,
}
```
to editor.edit?

AgentRBY replied:
Added

zardoy reviewed (changes requested):
I pushed some minor changes, so be sure to check them! Other things I leave to you for now! I also did some changes on main branch, so could also merge or rebase them?

zardoy commented:
Also what about case that I mentioned above:

```jsonc
{
    "key": "value" //comment
}
```

I think its still would be good to handle cases like this (insert comma after `"value"`). What do you think?

I think these requested changes are final, just need to figure out why CI checks don't run on this pull request? 🤔 Maybe you need to reopen the pr?

zardoy commented on src/extension.ts:
```diff
@@ -110,4 +114,91 @@ export const activate = () => {
     })
+
+    vscode.workspace.onDidChangeTextDocument(
+        ({ contentChanges, document }) => {
```
Also destructure `reason` here, and do check like in https://github.com/zardoy/vscode-better-snippets/blob/829a6bde0886a007e1cd7daf9a59ce850e601ea6/src/extension.ts#L314

zardoy commented on src/extension.ts:
```diff
@@ -110,4 +114,91 @@ export const activate = () => {
+                        continue;
+                    }
+
+                    if (document.languageId === 'jsonc') {
```
I think it would be better to remove this check, even if user still uses comments inside json (not jsonc) we still have to do comma placing at the right position.

AgentRBY committed:
Merge remote-tracking branch 'upstream/main'

AgentRBY committed:
Fix isNumber

AgentRBY committed:
Move condition to upperScope

AgentRBY committed:
Add support comments in classic JSON

AgentRBY committed:
Add checking reason

AgentRBY committed:
Add undoStopAfter and undoStopBefore

AgentRBY committed:
Add support Inserting comma before comment

AgentRBY commented:
> Also what about case that I mentioned above:
> 
> ```js
> {
>     "key": "value" //comment
> }
> ```
> 
> I think its still would be good to handle cases like this (insert comma after `"value"`). What do you think?

Done.

> I think these requested changes are final, just need to figure out why CI checks don't run on this pull request? 🤔 Maybe you need to reopen the pr?

Maybe, IDK. After all the issues are resolved, I will try to close this PR and create a new one

zardoy commented on src/extension.ts:
```diff
@@ -109,4 +114,94 @@ export const activate = () => {
+                        continue;
+                    }
+
+                    const textWithouComments = stripJsonComments(document.getText());
```
```suggestion
                    const textWithoutComments = stripJsonComments(document.getText());
```

zardoy replied:
Also let's move this out of the loop as we process absolutely the same text for each cursor position

AgentRBY replied:
I made some improvements so that the comment stripping function is called only once and only after all conditions to minimize its call

zardoy commented:
Tried to test this:

```jsonc
{
    "key": "value" //comment|
}
```

When pressing Enter at `|` position comma still doesn't get inserted after `"value"` e.g.:

```jsonc
{
    "key": "value", //comment
    |
}
```

You might want to get `prevLineLastChar` from `textWithoutComments`. Don't forget to pass `trailingCommas: false` as an option for `stripJsonComments`.

And I also don't understand the purpose of `isCurrentLineBeforeComment`.

AgentRBY committed:
Fix typo

Co-authored-by: Vitaly <vital2580@icloud.com>

AgentRBY committed:
Refactor and perfomance

AgentRBY committed:
Merge branch 'main' of https://github.com/AgentRBY/vscode-fix-all-json

AgentRBY commented:
> When pressing Enter at `|` position comma still doesn't get inserted after `"value"`

Oh, I misunderstood what you want to do a little, but I still implemented another good fix (see gif below)

> And I also don't understand the purpose of isCurrentLineBeforeComment.

![Code_4U5EbfK9UA](https://user-images.githubusercontent.com/68181944/203102136-d95a41af-780c-40ca-91e3-2ecd549c3063.gif)

AgentRBY committed:
Support inserting in end of comment

AgentRBY commented:
> When pressing Enter at `|` position comma still doesn't get inserted after `"value"` e.g.:

Also, I implemented your proposal

![Code_xSVs2edL9v](https://user-images.githubusercontent.com/68181944/203107591-a8d05def-5948-40ec-bf30-6fec8317f7d5.gif)

AgentRBY committed:
Minor fix

AgentRBY committed:
Replace trimEnd to trim

zardoy commented on src/extension.ts:
```diff
@@ -109,4 +114,105 @@ export const activate = () => {
+
+                    const isCurrentLineEmpty =
+                        currentLineText.trim() === "";
+                    const isCurrentLineBeforeComment = startsWithComment(currentLineText);
```
```suggestion
                    const isCurrentLineStartsWithComment = startsWithComment(currentLineText);
```

zardoy commented on src/extension.ts:
```diff
@@ -109,4 +114,105 @@ export const activate = () => {
+                        continue;
+                    }
+
+                    const fileContentWithoutComments = getFileContentWithoutComments();
```
I think it'd be better to replace `let temporaryFileWithoutComments: string` with `let fileWithoutComments: string` and do something like (no need for declaring function above)

```suggestion
                    fileWithoutComments ??= stripJsonComments(document.getText(), { whitespace: false });
```

zardoy replied:
FYI `??=` is shorthand for `var = var ?? ...`

Also don't undertsand the purpose of whitespace: false, would be better to trailingCommas: false instead

AgentRBY replied:
> I think it'd be better to replace let temporaryFileWithoutComments: string with let fileWithoutComments: string and do something like (no need for declaring function above)

Done

> Also don't undertsand the purpose of whitespace: false, would be better to trailingCommas: false instead

Already deleted it (also `trailingCommas: false` set by default)

zardoy commented on src/extension.ts:
```diff
@@ -109,4 +114,105 @@ export const activate = () => {
+
+                    const isComment = prevLineTextWithoutComments.trim() !== prevLineText.trim();
+
+                    const prevLineLastChar = isComment ? prevLineTextWithoutComments.trim().at(-1) : prevLineText.at(-1);
```
Why don't always grab last character from `prevLineTextWithoutComments`?

AgentRBY replied:
Did a little refactoring. Please test it

zardoy replied:
Yes, this is much better

zardoy commented on src/extension.ts:
```diff
@@ -109,4 +114,105 @@ export const activate = () => {
+                    }
+
+                    const fileContentWithoutComments = getFileContentWithoutComments();
+                    const prevLineTextWithoutComments = getTextByLine(fileContentWithoutComments, prevLine.lineNumber)!;
```
You need to remove whitespace: false, you can test that with current implementation and the following document:

```jsonc
{ /*test
[]|
*/
```
| - cursor pos to Enter
It'd throw runtime error (becuase you are using `!` but line actually doesn't exist)

AgentRBY replied:
Hmm, maybe fixed that. Please try to test again

AgentRBY committed:
Remove "whitespace: false"

AgentRBY committed:
Update src/extension.ts

Co-authored-by: Vitaly <vital2580@icloud.com>

AgentRBY committed:
Fix

AgentRBY committed:
Replcae to ??=

AgentRBY committed:
Refactoring

zardoy commented on src/extension.ts:
```diff
@@ -109,4 +114,97 @@ export const activate = () => {
+
+            void editor.edit((edit) => {
+                for (const [i, change] of contentChanges.entries()) {
+                    const prevLinePosition = change.range.start;
```
Let's inline this variable as the name is actually incorrect

zardoy commented:
I'll try to add & push integration tests soon (probably tomorrow)

AgentRBY committed:
Merge remote-tracking branch 'upstream/main'

AgentRBY committed:
Inline var

AgentRBY commented:
What do you think about the case when pasting string:
![Code_3am9I9UF3O](https://user-images.githubusercontent.com/68181944/203378893-05e96a53-e770-4b15-ad48-582be3e1e429.gif)
Do we have to insert a comma in this case?

zardoy commented on package.json:
```diff
@@ -67,6 +67,11 @@
+                "insertMissingCommaOnEnter": {
+                    "type": "boolean",
+                    "default": true,
+                    "description": "Insert missing comma after \"Enter\". Working only if last symbol is `}`, `]`, `\"` or number"
```
```suggestion
                    "description": "Insert missing comma after \"Enter\". Works only if last JSON symbol on the line is `}`, `]`, `\"` or number"
```

zardoy commented:
> Do we have to insert a comma in this case?

I don't think its a common case and let's not do this, because the setting description explicitly says that we do this only after `Enter`, so this behavior after pasting can be surprising & inconsistent to user. Futhermore user might not know wether just copied string starts with newline and I can imagine a more common case where user hits `Enter` and then pastes copied JSON content. That's why I was requesting that change above.

zardoy commented on src/extension.ts:
```diff
@@ -109,4 +114,95 @@ export const activate = () => {
         waitUntil(performFixes())
     })
+
+    vscode.workspace.onDidChangeTextDocument(
```
The feature became larger, let's extrat it to sibling file like `commaOnEnter.ts` (and just do `registerCommaOnEnter()` in extension.ts), also don't forget to run `pnpx prettier --write src/**`

AgentRBY committed:
Update package.json

Co-authored-by: Vitaly <vital2580@icloud.com>

AgentRBY committed:
Move to commaOnEnter and prettier all

AgentRBY committed:
Prettier

zardoy committed:
add comma on enter integration tests

zardoy committed:
fix tests

zardoy committed:
style: merge some if-return statements to reduce number of lines

zardoy committed:
Merge branch 'main' into pr/AgentRBY/6

zardoy committed:
update setting description (again)

zardoy commented:
Hey, @AgentRBY I've pushed integration tests and also fixes for them. Can you please wether they are good for you? I tried to to cover only basic cases, but probably there are any other cases that you would like to cover?

UPD: Also I removed the check that current line starts with comment. I didn't remember what cases it was covering? Gonna be honest, I don't think we ever need `startsWithComment` function here.

Meanwhile I'll try to fix CI issues here.

zardoy committed:
ci: try to delay tests

zardoy committed:
workaround vscode builtin JSON bug

zardoy committed:
Merge branch 'main' into pr/AgentRBY/6

zardoy committed:
test: avoid falsey triggers because of other test files

zardoy committed:
ci: try to ignore empty diagnostics update?

AgentRBY committed:
Add more tests

AgentRBY commented:
> Hey, @AgentRBY I've pushed integration tests and also fixes for them. Can you please wether they are good for you? I tried to to cover only basic cases, but probably there are any other cases that you would like to cover?

I've added some other simple tests to test more cases
Also, I think we should add tests for the pasting after comment (as in gif):
![Code_pJ2Ftmtsu3](https://user-images.githubusercontent.com/68181944/203648163-4829bdbd-02d1-4af2-b998-5e90be00aff1.gif)
But I haven't figured out how to do it, so I'll leave it to you, @zardoy 

> UPD: Also I removed the check that current line starts with comment. I didn't remember what cases it was covering? Gonna be honest, I don't think we ever need `startsWithComment` function here.

It was needed rather for optimization, in order to immediately discard values ​​that begin with comments. In general, its removal does not harm anything.

AgentRBY committed:
Change EoL checking

AgentRBY committed:
Format

zardoy commented:
> Also, I think we should add tests for the pasting after comment

I was going to add them as well, but forgot about, thanks for pointing!

zardoy commented on test/integration/suite/commaOnEnter.test.ts:
```diff
@@ -0,0 +1,92 @@
+        "key5": []/*|*/
+        "key6": {/*$*/
+        }/*|*/
+        "key7": "/*$*/ value",
```
Didn't get idea of this check, line below will be inserted at this position (string remains unchanged). Imagine that `ctrl+Enter` is being done on `/*$*/` marker (not just `Enter`). I will fix it as following (as in this position comma could be potentially inserted):
```suggestion
        "key7": "value",/*$*/
```

zardoy committed:
adjust tests

zardoy commented:
@AgentRBY I misread your comment:

> I've added some other simple tests to test more cases
> Also, I think we should add tests for the pasting after comment (as in gif):

I didn't get the idea about what you were talking, as I can see in the gif, you're just doing `ctrl+Enter` or Enter at the end. We already have test for this: https://github.com/zardoy/vscode-fix-all-json/pull/6/files#diff-cfd18f0b3c283dd23cf5513db00647d1949112658b52c7f406359cc92acdec00R19

zardoy reviewed (approved):
All LGTM! Can I merge, or you want anything else to add, @AgentRBY ?

AgentRBY commented:
> I didn't get the idea about what you were talking, as I can see in the gif, you're just doing ctrl+Enter or Enter at the end. We already have test for this: https://github.com/zardoy/vscode-fix-all-json/pull/6/files#diff-cfd18f0b3c283dd23cf5513db00647d1949112658b52c7f406359cc92acdec00R19

Okay, I misunderstood the test you pointed out a bit. I just checked and it's really what I wanted

I think it's ready to merge @zardoy

zardoy commented:
@AgentRBY Thank you so much for the idea & implementation 🎉. I'm happy that now I have solution when editing JSON in the web. Do you want to handle a similar problem: https://github.com/zardoy/vscode-fix-all-json/issues/9 ?

AgentRBY commented:
> @AgentRBY Thank you so much for the idea & implementation 🎉. I'm happy that now I have solution when editing JSON in the web. Do you want to handle a similar problem: #9 ?

Okay i will do it

Also check #10, if you like the idea, I will implement it too
